python/ray/tests/test_memory_deadlock.py

```diff
@@ -246,18 +236,119 @@ def test_deadlock_two_sets_of_task_with_nested_task(
-        parent_bytes, nested_bytes, barrier
+    nested_bytes = (
+        get_additional_bytes_to_reach_memory_usage_pct(memory_usage_threshold + 0.1)
+        - 2 * parent_bytes
     )
 
-    with pytest.raises(ray.exceptions.RayTaskError) as _:
-        ray.get(ref1)
-    with pytest.raises(ray.exceptions.RayTaskError) as _:
-        ray.get(ref2)
+    num_instances = 6
+    refs = [
+        task_with_nested_task.options(max_retries=-1).remote(parent_bytes, nested_bytes)
+        for _ in range(num_instances)
+    ]
+
+    if policy == LIFO_POLICY:
+        with pytest.raises(ray.exceptions.GetTimeoutError) as _:
+            for ref in refs:
+                ray.get(ref, timeout=120)
+    else:
+        for ref in refs:
+            ray.get(ref, timeout=120)
+
+
+@ray.remote
+def n_nested_tasks(n, task_bytes):
+    dummy = alloc_mem(task_bytes[0])
+    time.sleep(10)
+    if n > 1:
+        ray.get(n_nested_tasks.options(max_retries=-1).remote(n - 1, task_bytes[1:]))
+    return dummy[0]
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_two_sets_of_six_nested_tasks(
+    ray_with_memory_monitor,
+):
+    """task_with_nested_task allocates a block of memory, then runs
+    a nested task which also allocates a block memory. This is repeated
+    six times. This test runs two instances of this nesting."""
+
+    _, policy = ray_with_memory_monitor
+    parent_bytes = get_additional_bytes_to_reach_memory_usage_pct(
+        (memory_usage_threshold - 0.05) / 6
+    )
+    nested_bytes = (
+        get_additional_bytes_to_reach_memory_usage_pct(memory_usage_threshold + 0.1)
+        - 6 * parent_bytes
+    )
+
+    ref1 = n_nested_tasks.options(max_retries=-1).remote(
+        6, [parent_bytes] + [nested_bytes] * 5
+    )
+    ref2 = n_nested_tasks.options(max_retries=-1).remote(
+        6, [parent_bytes] + [nested_bytes] * 5
+    )
+
+    if policy == LIFO_POLICY:
+        with pytest.raises(ray.exceptions.GetTimeoutError) as _:
+            ray.get(ref1, timeout=300)
+            ray.get(ref2, timeout=300)
+    else:
+        ray.get(ref1, timeout=300)
+        ray.get(ref2, timeout=300)
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_churn_long_running(
+    ray_with_memory_monitor,
+):
+    long_running_bytes = get_additional_bytes_to_reach_memory_usage_pct(
+        memory_usage_threshold - 0.1
+    )
+    ray.get(
+        allocate_memory.options(max_retries=0).remote(
+            long_running_bytes, post_allocate_sleep_s=30
+        )
+    )
+    small_bytes = get_additional_bytes_to_reach_memory_usage_pct(
+        memory_usage_threshold + 0.2
+    )
+    with pytest.raises(ray.exceptions.GetTimeoutError) as _:
+        ray.get(allocate_memory.options(max_retries=-1).remote(small_bytes), timeout=45)
```

Comment:
probably need to add this

if __name__ == "__main__":
    sys.exit(pytest.main(["-sv", __file__]))